src/app/model/ChessEngine.ts

```diff
@@ -68,7 +70,78 @@ export class ChessEngine implements ChessModel {
     return side === Side.WHITE ? Side.BLACK : Side.WHITE;
   }
 
+  private returnWirtualProposedBoardAfterMove(chessboard: Chessboard, squareFrom: Square, squareTo: Square): Chessboard {
+    const wirtualPorposedChessboard = _.cloneDeep(chessboard);
+    const { squaresWithPiece: proposedSquaresWithPieces } = wirtualPorposedChessboard;
+
+    const movedPiece = chessboard.onPositionPiece(squareFrom);
+    const positionFrom = `${squareFrom.column}${squareFrom.row}`;
+    const positionTo = `${squareTo.column}${squareTo.row}`;
+    delete proposedSquaresWithPieces[positionFrom];
+    proposedSquaresWithPieces[positionTo] = movedPiece as Piece;
+
+    return wirtualPorposedChessboard;
+  }
+
+  public static kingPosition(chessboard: Chessboard, kingSide: Side): Square | undefined {
+    const { squaresWithPiece: squaresWithPieces } = chessboard;
+    const kingPositionKey = Object.keys(squaresWithPieces).find((key) => {
+      const isKingName = squaresWithPieces[key].name === 'King';
+      const isPlayerSide = squaresWithPieces[key].side === kingSide;
+      return isKingName && isPlayerSide;
+    });
+    const kingPosition = kingPositionKey
+      ? {
+          column: kingPositionKey[0],
+          row: Number(kingPositionKey[1]) as Row,
+        }
+      : undefined;
+
+    return kingPosition;
+  }
+
+  public static isSquareChecked(chessboard: Chessboard, playerSide: Side, positionToControl: Square): boolean {
+    let isCheckedSquareFlag = false;
+    const { squaresWithPiece: squaresWithPieces } = chessboard;
+
+    Object.keys(squaresWithPieces).forEach((key) => {
+      const mappedPiece = squaresWithPieces[key];
+      const isEnemySide = mappedPiece.side !== playerSide;
+
+      if (isEnemySide) {
+        const mappedPiecePosition = {
+          column: key[0],
+          row: Number(key[1]) as Row,
+        };
+        const possibleMappedPieceMoves = mappedPiece.possibleMoves(mappedPiecePosition, chessboard);
+        const isKingPositionOnPossibleEnemyPieceMoves = possibleMappedPieceMoves.some(
+          (checkedPosition) => JSON.stringify(checkedPosition) == JSON.stringify(positionToControl),
+        );
+        if (isKingPositionOnPossibleEnemyPieceMoves) isCheckedSquareFlag = true;
+      }
+    });
+    return isCheckedSquareFlag;
+  }
```

Comment:
Dlatego pytalem o programowanie funkcyjne :) Co powiesz na? 

```suggestion
  public static isSquareChecked(chessboard: Chessboard, playerSide: Side, positionToControl: Square): boolean {
    const squaresWithPieces = chessboard.squaresWithPiece;
    return Object.keys(chessboard.squaresWithPiece)
      .map(squareKey => ({position: {column: squareKey[0], row: Number(squareKey[1]) as Row}, piece: squaresWithPieces[squareKey]}))
      .filter(({piece}) => piece.side !== playerSide) //Skoro rozpatrujemy tylko przeciwnikow, tak sie mozemy pozbyc ifa. Moznaby to sprawdzanie dac do klasy Piece.
      .map(({position, piece}) => piece.possibleMoves(position, chessboard))
      .some(piecePossibleMoves => piecePossibleMoves.some(moveSquare => moveSquare.column === positionToControl.column && moveSquare.row === positionToControl.row))
  }
```

Reply:
Wiesz, że wczoraj przez godzinę nie mogłem usnąć, bo wpadłem na to, zeby tak zrobić i nie wiedziałem czy włączyć ponownie komputer czy iść spać?
A dziś nie mogłem sobie przypomnieć, jak to zrobić tak jak to co wczoraj mnie męczyło. Dzięki!!!!

Reply:
@nowakprojects CZARODZIEJ!!!

Reply:
Zrobiłeś to jeszcze lepiej niż sądziłem!